src/components/Combobox/components/ComboboxCore.js

```diff
@@ -0,0 +1,429 @@
+    /**
+     * When set to true, selecting an item will close the dropdown
+     */
+    closeOnSelect: PropTypes.bool,
+    comboboxClassName: PropTypes.string,
+    /**
+     * Number of results displayed in the result dropdown
+     */
+    height: PropTypes.oneOf([5, 7, 10]),
+    /**
+     * Unique identifier
+     */
+    id: PropTypes.string.isRequired,
+    /**
+     * Switch to render single selection as `inline` listboxs. Set by `EntityCombobox`
+     */
+    isInlineListboxSelection: PropTypes.bool,
+    /**
+     * Appends a `loading` spinner to the result items to
+     */
+    isLoading: PropTypes.bool,
+    /**
+     * Defines whether the Combobox allows more than one selection
+     */
+    isMultiSelect: PropTypes.bool,
+    /**
+     * Determines whether the result dropdown is rendered
+     */
+    isOpen: PropTypes.bool,
+    /**
+     * Array of items that will be displayed in the selection dropdown
+     * `isHeader` items break the dropdown into sublists
+     */
+    items: PropTypes.arrayOf(PropTypes.shape({
+      icon: PropTypes.shape({
+        icon: PropTypes.string.isRequired,
+        sprite: PropTypes.string.isRequired,
+      }),
+      id: PropTypes.string.isRequired,
+      isHeader: PropTypes.bool,
+      label: PropTypes.string.isRequired,
+      meta: PropTypes.node,
+    })),
+    /**
+     * Label of the Combobox
+     */
+    label: PropTypes.string.isRequired,
+    /**
+     * Assistive label for the Listbox of selected items
+     */
+    labelListbox: PropTypes.string,
+    /**
+     * Placeholder shown in the Dropdown input when no items are selected yet
+     */
+    placeholder: PropTypes.string.isRequired,
+    onSearch: PropTypes.func,
+    /**
+     * Callback to update selection state
+     * Called with (id, {
+     *   isReplace (bool) whether the passed `id` should replace the current selection state
+     *   isRemove (bool) whether the passed `id` is deselected
+     *   isAdd (bool) whether the passed `id` is an item not present in the Combobox
+     * })
+     */
+    onSelect: PropTypes.func.isRequired,
+    /**
+     * Callback to manage isOpen state
+     * Called with (bool) indicating next state
+     */
+    onToggle: PropTypes.func.isRequired,
+    /**
+     * Renders the input element
+     */
+    renderInput: PropTypes.func.isRequired,
+    /**
+     * Renders each non-header `item` in `items`
+     */
+    renderItem: PropTypes.func.isRequired,
+    /**
+     * Renders a `ComboboxDropdownItem` after the last displayed result
+     */
+    renderItemsAppended: PropTypes.func,
+    /**
+     * Renders a `ComboboxDropdownItem` after the last displayed result
+     */
+    renderItemsPrepended: PropTypes.func,
+    /**
+     * Renders beneath Combobox. Used to display a Listbox of Pills in multi-select Comboboxes
+     */
+    renderListbox: PropTypes.func,
+    /**
+     * Array of `id`s that are currently selected
+     */
+    selectedItems: PropTypes.arrayOf(PropTypes.shape({
```

Comment:
I implemented it as an array of id's at first but that leads to a set of problems which made me reconsider:

* Selectors to supply these are easier to write when both props accept the same shape. It's also easier to reason about
* I don't have to constantly use `.find()` on array `items` to lookup items when rendering the results which is an expensive operation on a large set of items
* This is a bit more flexible when it comes to implement custom variants in the future

Reply:
@quadreex I'm open to suggestions here though. There are at least two other approaches worth considering that allow cheap item lookups:

* Pass in `items` (keyed object), `resultItemIds` and `selectedItemIds`
* Pass in `resultItemIds` and `selectedItemIds` and lookup items via a prop function `getItem(id)`

I'm not 100% convinved that these are an improvement over the current solution when it comes to using the component but it's worth discussing all options here since this is the components "core"

Reply:
I'm a fan of clean and simple API.
For me, the best solution looks like `items` array prop and `selectedItemIds` array of selected IDs. I always like to store information in a most simple way (using most primitive types). So for me, string or number is a much more preferable holder of ID, than an object with several props.
Using an array of strings/numbers can potentially simplify redux logic (everything: selectors, reducers, action creators). Optimization of a component can lead to performance degradation of an app. For example, if I store selection as an array of IDs, every time component is rendered this array should be converted to an array of objects.

So generally I would prefer clean API over early optimizations.

As for complexity: If we take N for `items`'s length and M for 'selectedItemIds' from the first look it seems like a very complex algorithm (complexity NxM). But in real-world scenarios, both M and N rarely will be more than a 100 or 1000. And for edge cases in the future, we can optimize it by using additional keyed object internally.

I have not developed this component, and have not tested its performance, so maybe I'm missing something.

Reply:
You are quite right that the API would indeed be simpler when using item ids. There is however a case that would not be covered by passing `items` and `resultIds`:

A combobox can have a selection even though that selection is not part of `items`. This is the case when `selectedItems` contains a "prefilled" item or the user performs a search that returns different results while still preserving previous selections (when multiple selections are possible).

For that reason, a minimal solution having `selectedResultIds` would also need to have `searchResultIds` and a common `items` array / object.

Reply:
OK. Let's stick with the current implementation